Added parameters to get_relevant_area

dionytadema opened the pull request:
this action now supports the following parameters
**get_relevant_area(CallID, build, [amount, area, zones, stakeholder])**
the order of the parameters doesn't matter
the parameters are all optional
**amount(int)** // 15 if unspecified
**area(double, double)** //(200, 2000) if unspecified
**stakeholder(int)** // caller if unspecified
**zones(int, int, int....)** // all if unspecified

The functionality of this ticket is an extension on the work of WhySoSerious in #86 and implements #99 

[Edit by @MaxGroenenboom]: Travis: [![Build Status](https://travis-ci.org/levilime/tygron.svg?branch=customAzcAction)](https://travis-ci.org/levilime/tygron)

Commit:
added the action file, still trying to retrieve the parameters

Commit:
get_relevant_area_ext should now work

filters are now parsed, action should be
get_relevant_areas(CallID, Ext, [minsize, maxsize, [zones]]) 

renamed a variable to better distinguis between parameters and the given
filters

Commit:
added the get_relevant_area_ext to the action map

and reduced the amount of polygons to 1 to improve speed

Commit:
changed the min and max area ints to doubles

an area could be a double or float, we don't want the connector to crash
on this.

Commit:
removed cast to int

Commit:
removed cast altogether and used doubleValue()

Commit:
added a parameter to specify the max poly count

MaxGroenenboom commented:
Looks like a good addition, I do want to remark the following points:
- Why did you rename the parameters to filters? They even are used as parameters in your new action.
- Why did you change the integers in `GetRelevantAreasBuild` to doubles?
- This addition looks more like a modification of GetRelevantAreasBuild than a new action, and thus, it should have been added (by me) in GetRelevantAreasBuild.

I wasn't sure if I should add a system and parser for the filters as it would be restrictive and time consuming, but I suppose it would have been a good idea after all. The idea of the filters was having the ability to add functions to the parameterList, thus using your action in the end like this: 'get_relevant_areas(0, build, [max_polygons(#), min_size(#), max_size(#), zones(#,#)] with each # a number.

The idea of this action is good, it just should have been added in the build action (which I haven't because I ran out of time).

Would you mind rewriting it if I wrote a parser for parameters to `relevant_areas_action`?

dionytadema commented:
No that would be fine, and if we can also include the stakeholders that would be Nice,  why i renamed a variabele to filters is because in your bot you call the Action with get_relevant_areas(CallID, "build", filters) thus i named the variable in the Action to reflect that

I Just did't want to break your fuctionality

MaxGroenenboom commented:
Ah right, we used to call it filters, but renamed it to parameters to make it more generic :P
I'll start working on the parser now then, also in the case of filters you could just add them to the existing action, just make sure the filters are optional and the code still works if they aren't applied.

MaxGroenenboom commented:
Alright, I created a branch that makes handling parameters easier and more unified. I will create a PR as soon as Travis is done building.
https://github.com/Denpeer/tygron/tree/parameter_parser
Here is a comparison as well:
https://github.com/eishub/tygron/compare/context...Denpeer:parameter_parser

MaxGroenenboom commented:
A PR is now open for that which I mentioned in the last comments: #99

dionytadema commented:
Rewriting it to add the functionality to the get_relevant_area_build action would be fine by me, i'll work on it a little now, and then you can look at it and fix the parser

Commit:
added parameters to the get_relevant_area_build action

amount(int)				// 15 if unspecified
area(double, double)	//(200, 2000) if unspecified
stakeholder(int)		// caller if unspecified
zones(id, id , ...)		// all if unspecified

Commit:
fixed off by 1 error that resulted in one to many results

Commit:
Merge branch 'ExtendGetRelevantAreas' into customAzcAction

Conflicts:
	contextvh/src/main/java/contextvh/actions/GetRelevantAreas.java
	contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java

dionytadema commented:
updated, the functionality is now added to the already existing action

MaxGroenenboom commented:
Could you fix the checkstyle problems?
Also I would recommend handling the parameters like this:

```
int maxPolys = DEFAULT_MAX_POLYS;
if (parameters.containsKey("amount") {
    maxPolys = ((Numeral)parameters.get("amount").getFirst()).getValue().intValue();
}
```

Commit:
fixed checkstyle errors

MaxGroenenboom commented:
~~So for some reason the files changed overview in this PR shows a lot of changes that are already on the context branch,~~ (apparently this is by design, it shows a diff at the time the PR was created) while https://github.com/eishub/tygron/compare/context...levilime:customAzcAction doesn't show them. Don't know why that happens but to have a clearer overview of what this PR adds just follow that comparison link ^^

Commit:
fixed more checkstyle errors

Commit:
fixed last checkstyle errors?

MaxGroenenboom commented on contextvh/src/main/java/contextvh/actions/GetRelevantAreasBuild.java:
```diff
@@ -26,6 +28,10 @@
  */
 public class GetRelevantAreasBuild implements RelevantAreasAction {
 
+	private final int DEFAULT_MAX_POLYS = 10;
```
You should make these static or change the names, but statics would be the way to go.

MaxGroenenboom commented on contextvh/src/main/java/contextvh/actions/Parameters.java:
```diff
@@ -0,0 +1,44 @@
+	private static final long serialVersionUID = 6054823572206388545L;
+
+	/**.
+	 * Create a Empty Parameters object b
```
The dot here should be after the `b` of course ^^
what's that b doing there anyways?

EDIT checkstyle wanted it there, so i added it?

Commit:
added SuppressWarings to names of finals

dionytadema commented:
can someone fix checkstyle, i fixed all errors but it still fails on the name of my finals as they are all caps

Commit:
maybe checkstyle likes me now?

klikooo commented on contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:
```diff
@@ -73,7 +73,7 @@ public void testGetUsableLand() throws ManagementException, InterruptedException
 		//1m2 water
 		final double reservedArea = 9 + 4 + 1;
-		double area = action.getUsableArea(env.getEntity(MUNICIPALITY), null).getArea();
+		double area = action.getUsableArea(env.getEntity(MUNICIPALITY), new Parameters()).getArea();
```
No new test for a call to getUsuableArea with a zone filter?

dionytadema replied:
I'm working on this test now

dionytadema commented:
all checkstyle errors are fixed :smile:

TomBrunner commented:
Approved - CTD

spiderbiggen commented:
Approved once there are good tests with the different filters

-RAMpestampers

Commit:
added a test to test the zone filter

shdshdw commented:
I am not against the addition of this functionality. It seems that after the test has added that all issues from this PR now have been solved. Approved. 
- CodeFox [![CodeFox](https://avatars1.githubusercontent.com/u/18596181?v=3&s=32)](https://github.com/codefoxnl)

MaxGroenenboom commented on contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:
```diff
@@ -76,6 +77,27 @@ public void testGetUsableLand() throws ManagementException, InterruptedException
 	}
+	
+	/**
+	 * Test the getUsaubleLand function with zones
```
Typo in the javadoc (`Usauble` -> `Usable`)

MaxGroenenboom commented:
My eclipse is broken so I cannot test it at the moment, @denpeer is taking a look at it right now.

MaxGroenenboom commented:
Checkstyle fails for me locally on the new test, I don't know why travis doesn't.

ln 82, should end with a dot.
ln 95, magic number.
ln 99, whitespace before and after `/`.

Commit:
fixed some more checkstyle errors

Denpeer commented:
I still get a magic number error at 96 in `getRelevantAreasBuildTest`.

Denpeer commented:
otherwise it works fine.

MaxGroenenboom commented:
Take a look at option B in http://stackoverflow.com/a/30070188

Commit:
fixed last checkstyle error

jannelie commented:
We approve again -CTD

klikooo commented on contextvh/src/test/java/contextvh/actions/GetRelevantAreasBuildTest.java:
```diff
@@ -73,11 +74,34 @@ public void testGetUsableLand() throws ManagementException, InterruptedException
+	 * @throws InterruptedException {@link InterruptedException}
+	 */
+	@Test
+	@SuppressWarnings("checkstyle:magicnumber")
```
This is not necessary anymore?

Denpeer commented:
(comments deleted by @dionytadema) retrieved from slack github-bot
[12:40 PM]  
[eishub/tygron] New comment by Denpeer on pull request #98: Added parameters to get_relevant_area (assigned to dionytadema)
I still get a magic number error at 96 in `getRelevantAreasBuildTest`. 
It looks like you added @SuppressWarnings("checkstyle:magicnumber") but this isn't supported by checkstyle.

[12:40]  
[eishub/tygron] New comment by Denpeer on pull request #98: Added parameters to get_relevant_area (assigned to dionytadema)
otherwise it works fine. 

[12:44]  
[eishub/tygron] New comment by MaxGroenenboom on pull request #98: Added parameters to get_relevant_area (assigned to dionytadema)
Take a look at option B in http://stackoverflow.com/a/30070188

spiderbiggen commented:
@MaxGroenenboom it only fails locally because checkstyle on travis isn't setup to run on the tests.

MaxGroenenboom commented:
@spiderbiggen why isn't it? Or is it a bug? Either way, all of our code should be checkstyled, also tests.

spiderbiggen commented:
@MaxGroenenboom it defaults to have them not checked. see https://maven.apache.org/plugins/maven-checkstyle-plugin/faq.html#How_do_I_include_the_test_directory_in_Checkstyle for how to enable it

spiderbiggen commented:
Approved

-RAMpestampers

Denpeer commented:
Approved

-WhySoSerious
